sdlib/d/gc/thread.d

```diff
@@ -342,17 +344,20 @@ private:
 		return retry;
 	}
 
-	void scanSuspendedThreadsImpl(ScanDg scan) {
+	void scanThreadsImpl(ScanDg scan) {
 		assert(mThreadList.isHeld(), "Mutex not held!");
 
 		auto r = registeredThreads.range;
 		while (!r.empty) {
 			auto tc = r.front;
 			scope(success) r.popFront();
 
-			// If the thread isn't suspended, move on.
+			// We want to scan the current thread, and any
+			// suspended or detached threads. Other threads are
+			// likely GC helper threads, and don't need scanning.
 			auto ss = tc.state.suspendState;
-			if (ss != SuspendState.Suspended && ss != SuspendState.Detached) {
+			if (tc !is &threadCache && ss != SuspendState.Suspended
```

Comment:
IMO this is another indicator that the design is not quite right.

The function is being reworked to allow scanning the current thread in addition of the suspended threads, but if we are the current thread, we don't need to masquerade as a suspended thread, to call the suspended thread scan function, which have an exception to scan the current thread if not suspended. It could just call an appropriate function that does the scanning.